### PR TITLE
Implement infinite sums and products for lazy series

### PR DESCRIPTION
**Pull request by tscrim**
We allow lazy series to compute infinite sums `\sum_{i \in I} p_i` and products `\prod_{i \in I} (1 + p_i)` over arbitrary (enumerated) index sets `I` subject to a somewhat mild technical constraint that the order of each input `p_i` is weakly increasing wrt the iteration order and the preimage of each order is finite.

### :memo: Checklist

- [x] The title is concise, informative, and self-explanatory.
- [x] The description explains in detail what this PR is about.
- [x] I have linked a relevant issue or discussion.
- [x] I have created tests covering the changes.
- [x] I have updated the documentation accordingly.

### :hourglass: Dependencies

- #35265

### CODE REVIEW

**Comment by mantepse** [comment]
I will look properly tomorrow - maybe you could meanwhile merge develop?

One thing: essentially you only use `self._ring.one()`.  Maybe it would be better (and more in the spirit of `stream.py`) to just pass this, i.e., `one`?

Also, it would be good to include a check that it interacts nicely with recursive definitions.  Spoiler: looks good.
```
sage: f = L.undefined(valuation=0)
sage: f.define(L.prod(lambda n: t^n*f, PositiveIntegers()))
sage: f
1 + t + 2*t^2 + 5*t^3 + 12*t^4 + 30*t^5 + 77*t^6 + O(t^7)
sage: f[:10]
[1, 1, 2, 5, 12, 30, 77, 201, 532, 1427]
sage: oeis(f[:10])
0: A145267: G.f. satisfies A(x) = Product_{k>0} (1+x^k*A(x)).
```
In the docstring of `prod`, `index=True` should be `index_set=True`.

Did you check whether padics has similar functionality?

One thing that I do not really like is the dependency of the semantics on the input:
```
sage: L.prod([t^k for k in range(5)])
t^10
```
Wouldn't it be better to separate the two functionalities?  I don't know whether there is an established name for the infinite product of this specific form, though.

**Comment by mantepse** [comment]
Possibly `euler_product`?

Besides, here is possibly a bug:
```
sage: D = LazyDirichletSeriesRing(QQ, "s")
sage: D.prod(lambda p: (1+D(1, valuation=p)).inverse()-1, index_set=Primes())
```
does not return.

**Comment by tscrim** [comment]
I thought a lot about how to name it, and I couldn't come up with something natural beyond `prod()` (since it should also work for Dirichlet series). Ideally, it would be able to take in an infinite iterator and know to pass it off to the infinite product or just evaluate it out explicitly. However, that means we have to know whether an iterator is infinite or not, which is an impossible problem. Consequently I think of this `prod` as extending the current functionality, but if you feel it is better to separate the two, `euler_product` could work.

Indeed, that seems like a bug. It does "return" but it seems to have trouble getting coefficients. It probably has to do with not taking in the minimal valuation. This likely should be passed to the stream (which would be the default for `_cur_order`).

**Comment by tscrim** [comment]
Also, I did not check if padics has similar functionality. I agree with you about passing `one` instead of the ring. I thought I would use the ring more.

**Comment by mantepse** [comment]
Another (different) approach: wouldn't it be equally good to *require* that the factors passed are of the form $1+ f(x)$ with the valuation of $f$ increasing?  In the Dirichlet series example above, I have to manually subtract $1$ to get it into the required form.

I guess that your approach is slightly faster, but then we could have two methods, `prod` which just does the product and `euler_prod` which first adds one.

**Comment by tscrim** [comment]
> Another (different) approach: wouldn't it be equally good to _require_ that the factors passed are of the form $1+f(x)$ with the valuation of $f$ increasing? In the Dirichlet series example above, I have to manually subtract 1 to get it into the required form.

I thought about this, but I quickly came to the conclusion it is not a good idea. We would have to either manually implement something to check the next nonzero coefficient (and essentially duplicate our `valuation()` code) or subtract one and do a computation on something we are not actually going to use (which could also run into trouble with exact versus looks-like-1, say, with coefficients being lazy series). Indeed, it is not the "natural" input (and why you need to subtract `1`), but the implementation aspects suggest something better.
 
> I guess that your approach is slightly faster, but then we could have two methods, `prod` which just does the product and `euler_prod` which first adds one.

I also believe it to be faster. It is also better in terms of memory as the caches don't store all of those intermediate `0`'s. Moreover, I do not think the behaviors of the two methods should be different. In the proposal, the user is explicitly passing a different signature, so it is far less of a "gotcha" IMO.

TL;DR We can make it more natural by having $1 + f(x)$ as input, but there are tricky details with such an implementation.

**Comment by tscrim** [comment]
Perhaps it actually isn't so bad; I forgot that I did a custom check for the valuation on the components since I didn't need to compute their actual order.

**Comment by mantepse** [comment]
I could not find anything in padics.

I would still prefer to have a separate method for products of the form $\prod_i 1+f_i(x)$.  I do not see the benefit of having the same name for these.  It seems fitting to me to call this method `euler_product`, even though this has a much more specialised meaning in the context of Dirichlet series.

I agree that accepting the factors as input (i.e., including the 1) doesn't look very attractive, either.

**Comment by mantepse** [comment]
If it must be, and you make `prod` accept factors, that's OK.  I still think that having a second method would be a good idea, but I really don't want to argue about this.  I am only slightly worried that at some point we could have symbolic infinite products and lazy infinite products mean (slightly) different things.

Currently sage doesn't seem to have any infinite product at all.

**Comment by tscrim** [comment]
I changed it to have the iterator output $p_i(x) = 1 + f_i(x)$ and fixed the Dirichlet series bug (which was not limited to those series actually). However, in order to have these factors, I had to remove a safeguard against products involving smaller order $f_i(x)$ as I did not think it was useful to force the computation of all of those coefficients. We are already having to assume a fair amount of "good" user input for this.

To get rid of the infinite loop trap, I can change it so the order is only increased one step at each time. Although for something more "sparse" in regards to the orders of $f_i(x)$, this will mean multiplying more factors sooner, which can get very computationally expensive I think.

Anyways, that's all for me tonight; I need some sleep.

File: src/sage/data_structures/stream.py [hunk 1]
```diff
@@ -3220,3 +3220,144 @@ def is_nonzero(self):
             True
         """
         return self._series.is_nonzero()
+
+class Stream_infinite_product(Stream):
+    r"""
+    Stream defined by an infinite product.
+
+    The ``iterator`` returns elements either `1 + p_i` to compute the
+    product `\prod_{i \in I} (1 + p_i)`. The valuation of `p_i` is weakly
```
Comment (mantepse) [inline src/sage/data_structures/stream.py]: "either" does not fit here.

File: src/sage/data_structures/stream.py [hunk 2]
```diff
@@ -3220,3 +3220,144 @@ def is_nonzero(self):
...
+        """
+        if self._cur is None:
+            temp = next(self._prod_iter)
+            self._cur = temp
+            self._cur_order = temp._coeff_stream._approximate_order
+        order = self._cur_order
+        while order == self._cur_order:
+            try:
+                next_factor = next(self._prod_iter)
+            except StopIteration:
```
Comment (mantepse) [inline src/sage/data_structures/stream.py:3370]: We can assume here that the iterator never raises a `StopIteration`.

File: src/sage/data_structures/stream.py (hunk 1 above)
Comment (tscrim) [reply src/sage/data_structures/stream.py]: Leftovers. Fixed in my next commit.

File: src/sage/data_structures/stream.py (hunk 2 above)
Comment (tscrim) [reply src/sage/data_structures/stream.py:3370]: In principle, we should, but I don't want to prohibit it since it is a insignificant speed difference and the code should handle it (although I haven't tested it yet, much less added a doctest).

Comment (mantepse) [reply src/sage/data_structures/stream.py:3370]: For me it is not about speed but about clarity of the code.  If my future self some  years from now reads this code, it will have to think again whether this can actually happen.  However, it seems now that `index_set=True` with a finite generator that satisfies the valuation condition returns the finite product, which may not be so bad.

I admit that I am not completely sure what we want.  For example, having `index_set=Primes()` looks nice, but it is unclear to me whether it is really worth it.

I'd feel much better to have `prod` only for finite products and `euler_prod` or `infinite_prod` or `prod_inifinite` or whatever for potentially infinite products of the form $\prod_i (1 + f_i)$, where the valuation of `f_i` is weakly increasing with `i`, and get rid of `index_set` entirely.
```py
sage: L.<t> = LazyLaurentSeriesRing(QQ)
sage: L.prod((1 - t^n for n in range(1,3)), index_set=True)
1 - t - t^2 + t^3 + O(t^7)
sage: L.prod((1 - t^(10-n) for n in range(1,3)), index_set=True)
1 + O(t^7)
sage: L.prod((1 - t^(10-n) for n in range(1,3)))
1 - t^8 - t^9 + t^17
```

Comment (tscrim) [reply src/sage/data_structures/stream.py:3370]: I fully disagree with you. It is unnatural to not use `prod` for every type of product. Having to do something different for infinite products would _constantly_ annoy me. Having functions behave differently with different input is a normal thing for code; although it only requires different input because of a technical (well...in some sense mathematical) limitation. We can't get around undecidable problems either.

Admittedly I am not completely happy with `index_set=True` rather than, e.g., `infinite=True`, but it is convenient both in the code (no need to resolve mixed inputs) and as a user (only need to remember one argument and it behaves consistently with finite versus infinite iterators).

Note that the `index_set` might also be a (large-but-)finite set too (say, `index_set=range(1,3)`). However, if you are doing a finite product, the user should be passing a normal iterable anyways. The documentation clearly specifies what input is required for `index_set`-based input; i.e., garbage-in-garbage-out.

In an ideal world, we would be able to peer inside the iterator and be able to check that the infinite iterating object is not known to be finite. However, Python doesn't let us check that.

Comment (tscrim) [reply src/sage/data_structures/stream.py:3370]: One thing that could be done: If the `index_set` is known to be a finite set, then it just runs the product normally (basically, it does the job the user should be doing of putting it as a generator object).

Comment (mantepse) [reply src/sage/data_structures/stream.py:3370]: I just noticed that there is a significant speed penalty for having the factors being of the form $1+f_i$.  For example, starting from your second commit with the following minor modification:
```diff
diff --git a/src/sage/data_structures/stream.py b/src/sage/data_structures/stream.py
index 652ba51b93b..9f7a54b0723 100644
--- a/src/sage/data_structures/stream.py
+++ b/src/sage/data_structures/stream.py
@@ -3285,8 +3285,8 @@ class Stream_infinite_product(Stream):
                 if coeff_stream[coeff_stream._approximate_order]:
                     order = coeff_stream._approximate_order
                     raise ValueError(f"invalid product computation with invalid order {order} < {self._cur_order}")
-            self._cur *= self._one + next_factor
-            if not coeff_stream[coeff_stream._approximate_order]:
+            self._cur += self._cur * next_factor
+            while not coeff_stream[order]:
                 order += 1
         self._cur_order = order
```
I get the timing
```
sage: L.<t> = LazyLaurentSeriesRing(QQ)
sage: euler = L.prod(lambda n: -t^n, PositiveIntegers())
sage: %time euler[100]
CPU times: user 957 ms, sys: 3.17 ms, total: 961 ms
Wall time: 961 ms
1
```
vs.
```
sage: L.<t> = LazyLaurentSeriesRing(QQ)
sage: euler = L.prod(lambda n: 1 - t^n, PositiveIntegers())
sage: %time euler[100]
CPU times: user 8.62 s, sys: 0 ns, total: 8.62 s
Wall time: 8.62 s
1
```

Comment (mantepse) [reply src/sage/data_structures/stream.py:3370]: Maybe we should ask on sage-devel for other people whether they have an opinion?

Comment (tscrim) [reply src/sage/data_structures/stream.py:3370]: That's a good catch. It's likely quite possible to get similar speeds by doing `self._cur += self._cur * (next_factor - one)` with $1 + f_i$ factors.

If we are at an impasse, then we should solicit other opinions to break the tie.

Comment (mantepse) [reply src/sage/data_structures/stream.py:3370]: I tried to optimize the `1 + f_i` version, as you suggested.  This indeed makes it close in efficiency to the `f_i` version, although it is still a bit slower, maybe 10% or so.  However, I did not try hard to optimize either.  A simple `%lprun` indicates that almost all the time is spent in
```py
self._cur += self._cur * next_factor
```
which is no surprise.

Comment (mantepse) [reply src/sage/data_structures/stream.py:3370]: I think that for sums there is no question on how to provide the functionality, is there?

Comment (tscrim) [reply src/sage/data_structures/stream.py:3370]: We still need to have special arguments added to distinguish infinite sums versus finite ones. Granted, there is no yield from $1 + f_i$ versus $f_i$ issue. However, I thought your issue was more fundamental in that `prod` should not be taking other types of inputs/arguments? If it is just that it did not take functions/iterators that yield $1 + f_i$, then perhaps we are both okay with the current branch (modulo your other comments).

Comment (mantepse) [reply src/sage/data_structures/stream.py:3370]: My issue is that it's unclear to me what's best.  Let me summarize:

1. I am strictly against `prod` accepting `f(i)`.
2. I think that `prod((1+f(i) for i in I), True)` is very ugly and not self-explanatory.
3. I think that `euler_prod(it)` and `euler_prod(f, I)` is fine, and so is `prod(lambda i: 1 + f(i), I)`.
4. I think that we should use `sum` for sums of infinitely many series.
5. I think it is quite likely that creating `1 + f(i)` may lead to an unnecessary slowdown: for every factor, you have to compute `(1 + f(i)) - 1` and then detect the order, which involves skipping all the zeros.  Possibly one can get around this somehow, but it seems to me like a very unnecessary complication.

The ability to specify the index set is very nice.  Still, let us consider the alternative `prod(f, a, b)`:
```py
sage: L.prod(lambda n: x^Primes().unrank(n), 0, oo)
```
I admit that this does not look any better to me. 

By way of example:
```py
sage: L = LazyDirichletSeriesRing(QQ, "s")

sage: L.prod(lambda n: 1 - L([1], valuation=n), Primes())
sage: L.euler_prod(lambda n: - L([1], valuation=n), Primes())
sage: L.sum(lambda n: L([1], valuation=n), Primes())
```
looks good to me.  I dislike
```py
sage: L.prod((1 - L([1], valuation=n) for n in Primes()), True)
sage: L.sum((L([1], valuation=n) for n in Primes()), True)
```
Personally, I would maybe remove the possibility of summing / multiplying over an infinite iterator.  Do you have a use case for the infinite iterator, where it is more natural than providing a function and an index set?

Comment (tscrim) [reply src/sage/data_structures/stream.py:3370]: >     1. I am strictly against `prod` accepting `f(i)`.

I concur if there are no additional arguments (mainly a flag specifying $1 + f_i$ versus $f_i$ input).

>     2. I think that `prod((1+f(i) for i in I), True)` is very ugly and not self-explanatory.

Agreed, but it is a technical limitation we have to work around.

>     3. I think that `euler_prod(it)` and `euler_prod(f, I)` is fine, and so is `prod(lambda i: 1 + f(i), I)`.
> 
>     4. I think that we should use `sum` for sums of infinitely many series.

Infinite sums and products should follow the same paradigm. So if we use `sum` for infinite series, we should use `prod` for infinite products.

>     5. I think it is quite likely that creating `1 + f(i)` may lead to an unnecessary slowdown: for every factor, you have to compute `(1 + f(i)) - 1` and then detect the order, which involves skipping all the zeros.  Possibly one can get around this somehow, but it seems to me like a very unnecessary complication.

Depends on how much we want to optimize. At a certain point, one needs to write more complicated code to get faster speed. It would depend on how necessary it is.

> The ability to specify the index set is very nice. Still, let us consider the alternative `prod(f, a, b)`:
> 
> ```python
> sage: L.prod(lambda n: x^Primes().unrank(n), 0, oo)
> ```
> 
> I admit that this does not look any better to me.

Agreed, plus that would make the code even more complicated.

> Personally, I would maybe remove the possibility of summing / multiplying over an infinite iterator. Do you have a use case for the infinite iterator, where it is more natural than providing a function and an index set?

The code simply constructs the iterator and then passes that off. So an iterator is the "natural" object in terms of the code. It's also what the input should be anyways in an ideal world. Mainly, if we could probe the iterator to verify finiteness, then I would only take an iterator as input.

A note about generator inputs: My initial implementation was processing inputs analogous to what `map` does: `map(f, X)` instead of `(f(x) for x in X)`. Although it made for a convenient way for the user to say "treat this like an infinite product".

One way an iterator would be very natural is if you implemented some generator function `foo`  (so it uses `yield`) that doesn't have a natural indexing set and/or defined recursively. (Of course, a recursive function would likely have an index set of say, `NN`, but it would be very computationally inefficient to not have it be a generator.)

Comment (mantepse) [reply src/sage/data_structures/stream.py:3370]: Speed is of fundamental importance to me.  In a past project (about guessing formulas, using FriCAS), it turned out that the series code was the bottleneck (which in fact made some computations impossible), and this is very hard to optimize quickly.  So I won't give up on having `Stream_infinite_product` being based on the `f(i)`.  There shouldn't be a problem having two related, but different methods.

The more I think about it, the more I would abandon allowing infinite generators together with `index_set=True`.  If, for some reason, I already have the iterator `it`, I can turn it into a function mapping from `NN` without loss of performance, but I doubt that this will happen frequently:
```
sage: f = lazy_list(it)
sage: L.prod(lambda n: f[n], NN)
```

Comment (tscrim) [reply src/sage/data_structures/stream.py:3370]: I am remain convinced that we should fully support infinite generators. They are too natural to not include as I have previously mentioned; let me stress that the code itself converts the input to a generator. As such, I think it is confusing both for a user and implementor to have two entry points for the same operation (even though the inputs formats are slightly different).

The `lazy_list` is a nice way around passing a generator object. Yet you are now storing outputs that you don't need, and they are then just fed back into a generator. It has a code smell to me. (Plus it is not in the global namespace by default, so it would need a very knowledgeable user.)

Now I would be okay with having the separate method that only takes the $f_i$ as the inputs (both as a generator and/or function&index), but I highly doubt I will be convinced moving off having `prod()` take the same input (differentiating from $1 + f_i$ input through an additional boolean argument).

Comment (mantepse) [reply src/sage/data_structures/stream.py:3370]: I don't understand what you mean with your last paragraph.

Comment (mantepse) [reply src/sage/data_structures/stream.py:3370]: Two minor things: yes, ·lazy_list· should be in the global namespace. I use it almost every day.

I don't think that using the lazy list no additional stuff I'd kept in memory. All the series in the iterator must be kept in memory, the lazy list elements are just pointers to them.

Comment (tscrim) [reply src/sage/data_structures/stream.py:3370]: We can have `infinite_prod()` or whatever take inputs that yield $f_i$ for the product $\prod_{i \in I} 1 + f_i$, but I also want `prod()` to do that as well. I think it makes sense to have both take an additional flag, say `add_one`, that differentiates between the iterator yielding $f_i$ or $1 + f_i$.

> I don't think that using the lazy list no additional stuff I'd kept in memory. All the series in the iterator must be kept in memory, the lazy list elements are just pointers to them.

That is not true. You do not need to keep the series $f_i$ (or $1 + f_i$) in memory once they are added to the product. In particular, this happens for eventually constant series.

**Comment by tscrim** [comment]
I added the doctests and fixed a few small things (it handles trivial zeros in the iterator). This is now ready for review.

**Comment by mantepse** [comment]
I discovered a problem, which I don't understand yet:
```
L.<z> = LazyPowerSeriesRing(QQ)
T = L.undefined(1)
D = L.undefined(0)
T.define(z*exp(T)*D)
H = L(lambda n: sum(T(z^k)/k for k in range(2, n+1))[n], valuation=2)
D.define(exp(H))
```
works, but
```
L.<z> = LazyPowerSeriesRing(QQ)
T2 = L.undefined(1)
D2 = L.undefined(0)
T2.define(z*exp(T2)*D2)
H2 = L.sum(lambda k: T2(z^k)/k, 2)
D2.define(exp(H2))
```
raises a `StopIteration`.  It might be a user error, of course, but I don't see it yet.

**Comment by tscrim** [comment]
That is a good example. It shows that my implementation was computing coefficients when it shouldn't have been. I have changed things around so that it computes fewer coefficients unless it absolutely needs to. The drawback is that we need to be a bit more careful with having the approximate order weakly increasing when using undefined series in infinite sums/products.

**Review by mantepse — APPROVED** [review]
Looks good!

**Comment by tscrim** [comment]
Thank you.
